schema-engine/sql-schema-describer/src/postgres/constraints_query.sql

```diff
@@ -0,0 +1,16 @@
+SELECT
+    schemainfo.nspname AS namespace,
+    tableinfo.relname AS table_name,
+	constr.conname AS constraint_name,
+	constr.contype AS constraint_type,
+	pg_get_constraintdef(constr.oid) AS constraint_definition,
+	constr.condeferrable AS is_deferrable,
+	constr.condeferred AS is_deferred
+FROM pg_constraint constr
+JOIN pg_class AS tableinfo
+	ON tableinfo.oid = constr.conrelid
+JOIN pg_namespace AS schemainfo
+	ON schemainfo.oid = tableinfo.relnamespace
+WHERE schemainfo.nspname = ANY ( $1 )
+	AND contype NOT IN ('p', 'u', 'f')
```

Comment:
This excludes:
- `p`: `PRIMARY KEY` constraints
- `u`: `UNIQUE` constraints, which are retrieved via indexes
- `f`: `FOREIGN KEY` constraints

Reply:
Is there an obstacle to querying this in the same query as foreign keys (I think it's the same tables as the existing foreign keys query)? That would save a roundtrip to the database for every call to describe, and we do call describe a lot in migrate.

Reply:
The major obstacle is that foreign keys and check/exclusion constraints require different info, and use very different JOINs.

How about:

```sql
WITH cte_constraints AS (
	SELECT
		constr.oid AS constraint_id,
		constr.conname AS constraint_name,
		constr.contype AS constraint_type,
		pg_get_constraintdef(constr.oid) AS constraint_definition
	FROM pg_constraint constr
	JOIN pg_class AS tableinfo
		ON tableinfo.oid = constr.conrelid
	JOIN pg_namespace AS ns
		ON ns.oid = tableinfo.relnamespace
    WHERE
      ns.nspname = ANY (ARRAY['public']) -- ANY $1
      AND constr.contype IN ('f', 'x', 'c')
)
SELECT json_build_object(
	'check',
		(
	        SELECT json_agg(json_build_object(
				'constraint_name', constr.conname,
				'constraint_definition', pg_get_constraintdef(constr.oid),
				'is_deferrable', constr.condeferrable,
				'is_deferred', constr.condeferred
	        ))
	        FROM cte_constraints ctec
	        JOIN pg_constraint constr
	        	ON constr.oid = ctec.constraint_id
	        JOIN pg_class
	        	ON constr.conrelid = pg_class.oid
	        WHERE contype = 'c'
		),
	'foreign_keys',
		(
			SELECT json_agg(json_build_object(
				'con_id', con.oid,
				'child_column', att2.attname,
				'parent_table', cl.relname,
				'parent_column', att.attname,
				'confdeltype', con.confdeltype,
				'confupdtype', con.confupdtype,
				'referenced_schema_name', rel_ns.nspname,
				'constraint_name', conname,
				'child', child,
				'parent', parent,
				'table_name', table_name,
				'namespace', namespace
			) ORDER BY con.namespace, table_name, conname, con.oid, con.colidx)
			FROM (
				SELECT
                        ns.nspname AS "namespace",
                        con1.contype,
                        unnest(con1.conkey)                AS "parent",
                        unnest(con1.confkey)                AS "child",
                        cl.relname                          AS table_name,
                        ns.nspname                          AS schema_name,
                        generate_subscripts(con1.conkey, 1) AS colidx,
                        con1.oid,
                        con1.confrelid,
                        con1.conrelid,
                        con1.conname,
                        con1.confdeltype,
                        con1.confupdtype
		        FROM cte_constraints ctec
		        JOIN pg_constraint con1
		        	ON con1.oid = ctec.constraint_id
		        JOIN pg_class cl
		        	ON con1.conrelid = cl.oid
		        JOIN pg_namespace ns
		        	ON cl.relnamespace = ns.oid
                WHERE con1.contype IN ('f', 'x', 'c')
                ORDER BY colidx
			) con
                    JOIN pg_attribute att on att.attrelid = con.confrelid and att.attnum = con.child
                    JOIN pg_class cl on cl.oid = con.confrelid
                    JOIN pg_attribute att2 on att2.attrelid = con.conrelid and att2.attnum = con.parent
                    JOIN pg_class rel_cl on con.confrelid = rel_cl.oid
                    JOIN pg_namespace rel_ns on rel_cl.relnamespace = rel_ns.oid
           )
	) as constraints
```

Reply:
> This type of queries requires rewriting parts of the logic for get_foreign_keys() and get_constraints() for Postgres

get_constraints is new so it's fine, and get_foreign_keys should not get more complicated, so I think that's fine :+1:

> The same structure for this query may not be transposable to MySQL and other databases with weaker json-aggregation support

I also think that's fine, we'll deal with that when we get there, the describers are completely separated by database and we don't make effort to sync them in that way.

---

I like the approach and I think that should work, and that addresses the extra roundtrips, let's give it a try!

---

I haven't read the query in very exhaustive details, but in the foreign keys query, it looks like we're joining twice on pg_class with the same condition (lines with the aliases `rel_cl` and `cl`).

---

re: separating the foreign keys key and the check key because different shape: again I haven't looked super deep, but it feels like they could be the same rows, with nullable columns depending on `contype` — the joins in the foreign keys query would have to become outer joins, but beyond that I'm not sure you'd have to change much. This is an impression that we could maybe do simpler, not a strongly held opinion, so please feel free to ignore.

Reply:
I agree we could load these together with foreign keys.